mailpoet/lib/Automation/Engine/Data/WorkflowRunLog.php

```diff
@@ -0,0 +1,174 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Automation\Engine\Data;
+
+use DateTimeImmutable;
+use InvalidArgumentException;
+use MailPoet\Automation\Engine\Exceptions\InvalidStateException;
+use MailPoet\Automation\Engine\Exceptions\UnexpectedValueException;
+use MailPoet\Automation\Engine\Utils\Json;
+use Throwable;
+
+class WorkflowRunLog {
+
+  const STATUS_RUNNING = 'running';
+  const STATUS_COMPLETED = 'completed';
+  const STATUS_FAILED = 'failed';
+
+  /** @var int */
+  private $id;
+
+  /** @var int */
+  private $workflowRunId;
+
+  /** @var DateTimeImmutable */
+  private $startedAt;
+
+  /** @var DateTimeImmutable|null */
+  private $completedAt;
+
+  /** @var string */
+  private $status;
+
+  /** @var array */
+  private $error;
+
+  /** @var array */
+  private $data;
+
+  /** @var string */
+  private $stepId;
+
+  /** @var array */
+  private $args;
+
+  public function __construct(
+    int $workflowRunId,
+    string $stepId,
+    array $args,
+    int $id = null
+  ) {
+    $this->workflowRunId = $workflowRunId;
+    $this->stepId = $stepId;
+    $this->args = $args;
+    $this->status = self::STATUS_RUNNING;
+
+    if ($id) {
+      $this->id = $id;
+    }
+
+    $this->startedAt = new DateTimeImmutable();
+
+    $this->error = [];
+    $this->data = [];
+  }
+
+  public function getId(): int {
+    return $this->id;
+  }
+
+  public function getWorkflowRunId(): int {
+    return $this->workflowRunId;
+  }
+
+  public function getStepId(): string {
+    return $this->stepId;
+  }
+
+  public function getStatus(): string {
+    return $this->status;
+  }
+
+  public function getArgs(): array {
+    return $this->args;
+  }
+
+  public function getError(): array {
+    return $this->error;
+  }
+
+  public function getData(): array {
+    return $this->data;
+  }
+
+  /**
+   * @return DateTimeImmutable|null
+   */
+  public function getCompletedAt() {
+    return $this->completedAt;
+  }
+
+  /**
+   * @param string $key
+   * @param mixed $value
+   * @return void
+   */
+  public function setData(string $key, $value): void {
+    try {
+      $newData = $this->getData();
+      $newData[$key] = $value;
+      $encoded = Json::encode($newData);
+      $decoded = Json::decode($encoded);
+      if ($decoded !== $newData) {
+        throw new InvalidArgumentException('$value must be serializable');
+      }
+    } catch (InvalidStateException | InvalidArgumentException | UnexpectedValueException $e) {
+      throw new InvalidArgumentException("Invalid data provided for key $key.");
+    }
```

Comment:
This is probably not the best way to handle this, but I do think it's important that we don't allow setting vales that we wouldn't be able to serialize/JSON encode and retrieve out of the database. Is there a more elegant way of handling this?

Reply:
What is it supposed to do, exactly? The `Json::encode()` function will throw a `InvalidStateException` if the input is not serializable, so that could be enough by itself, couldn't it? If you want to convert it to `InvalidArgumentException`, we can keep the catch + throw.

It may be also a bit confusing that you throw `new InvalidArgumentException('$value must be serializable')` and immediately below catch it, using a different message. As for `UnexpectedValueException`, that's not even thrown anywhere, is it?

I also just realized this is not used anywhere yet, but I guess it makes sense to have it.

Reply:
I agree that this code is very confusing and I need to make it simpler.

> I also just realized this is not used anywhere yet, but I guess it makes sense to have it.

The main idea here is that integrations would be able to use this public method to store arbitrary data in the log, since the log object is getting exposed in a WordPress action.

The general idea with the check is I don't want to allow setting any kind of data that would end up being different after it's retrieved from the database and decoded.

Taking an example from one of the tests I wrote, you could have something like this:

```php
$badData = [
  function() { echo 'closures cannot be serialized'; }
];
$log->setData('badData', $badData);
```

This doesn't cause an error when given to `Json::encode`, but the encoded version just ends up being an empty object:

```json
{"badData":[{}]}
```

Trying to JSON encode a newsletter object causes a similar result (encoded without error as an empty object). And then there are cases where someone might try to store something like a stdClass, which works very well with encoding, but after decoding it it would be an array instead of a stdClass, which could be confusing.

If we wanted to allow storing more complex data, like objects, we could use PHP serialization instead of JSON, but that seems unnecessary and more likely to cause problems than anything.

Right now I'm leaning towards limiting the type of $value to just scalars, maybe including arrays that only have scalar values.

`UnexpectedValueException` would come from the [jsonNotObject](https://github.com/mailpoet/mailpoet/blob/MAILPOET-4463%2Fadd-step-logging/mailpoet/lib/Automation/Engine/Utils/Json.php#L25-L25) exception from decoding the JSON. I'm not sure if it would even be possible to trigger that in this case (seems doubtful) but I wanted to be exhaustive with the list of exceptions I was catching.

Reply:
I went ahead with allowing scalar values and arrays that exclusively contain scalar values. If this seems overly complex then I'm happy to change it.